src/main/java/com/flippingutilities/ui/widgets/SlotStateDrawer.java

```diff
@@ -0,0 +1,521 @@
+            existingQuickLookWidget.setHidden(false);
+        }
+    }
+
+    /**
+     * This is the text widget that shows how long the offer has been competitive since
+     * the last time it was updated.
+     */
+    private void addTimeInRangeWidget(Widget slotWidget, int slotIdx, long latestTimeInRange) {
+        Widget existingWidget = slotIdxToTimeInRangeWidget.get(slotIdx);
+        if (existingWidget == null || !isWidgetStillAttached(existingWidget)) {
+            Widget timeInRangeWidget = createTimeInRangeWidget(slotWidget);
+            timeInRangeWidget.setText(TimeFormatters.formatDuration(Duration.ofSeconds(latestTimeInRange)));
+            slotIdxToTimeInRangeWidget.put(slotIdx, timeInRangeWidget);
+        } else {
+            existingWidget.setHidden(false);
+            existingWidget.setText(TimeFormatters.formatDuration(Duration.ofSeconds(latestTimeInRange)));
+        }
+    }
+
+    private Widget createQuicklookWidget(Widget slotWidget, RemoteSlot slot) {
+        Widget quickLookWidget = slotWidget.createChild(-1, WidgetType.GRAPHIC);
+        quickLookWidget.setFontId(FontID.PLAIN_11);
+        quickLookWidget.setYPositionMode(WidgetPositionMode.ABSOLUTE_TOP);
+        quickLookWidget.setOriginalX(90);
+        quickLookWidget.setOriginalY(52);
+        quickLookWidget.setSpriteId(SpriteID.BANK_SEARCH);
+        quickLookWidget.setWidthMode(WidgetSizeMode.ABSOLUTE);
+        quickLookWidget.setOriginalHeight(22);
+        quickLookWidget.setOriginalWidth(22);
+        quickLookWidget.setXPositionMode(WidgetPositionMode.ABSOLUTE_LEFT);
+        quickLookWidget.setXTextAlignment(WidgetTextAlignment.LEFT);
+        quickLookWidget.setTextShadowed(true);
+        quickLookWidget.setHasListener(true);
+
+        quickLookWidget.setOnMouseOverListener((JavaScriptCallback) ev -> {
+            SwingUtilities.invokeLater(() -> {
+                if (this.wikiRequest == null) {
+                    quickLookPanel.updateDetails(null, null);
+                    return;
+                }
+                Optional<RemoteSlot> maybeRemoteSlot = unifiedSlotStates.get(slot.getIndex());
+                if (!maybeRemoteSlot.isPresent()) {
+                    quickLookPanel.updateDetails(null, null);
+                    return;
+                }
+
+                int itemId = maybeRemoteSlot.get().getItemId();
+                WikiItemMargins margins = this.wikiRequest.getData().get(itemId);
+                PointerInfo a = MouseInfo.getPointerInfo();
+                Point p = a.getLocation();
+                quickLookPanel.updateDetails(maybeRemoteSlot.get(), margins);
+                popup.pack();
+                popup.setLocation(p.x + 25, p.y);
+                popup.setVisible(true);
+            });
+        });
+
+        quickLookWidget.setOnMouseLeaveListener((JavaScriptCallback) ev -> {
+            SwingUtilities.invokeLater(() -> {
+                popup.setVisible(false);
+            });
+        });
```

Comment:
potential memory leak if the event listener is never deleted?
Idk if it would be caught by the GC, as the popup is never destroyed only hidden

Reply:
We recreate the quick look widget when it gets unattached from its parent (no longer shows up on the screen). The only reference to the widget object is in a hash map. When we recreate the widget, we remove the previous one from the hash map and put the newly created widget in its place. Since it the old widget has no references to it any longer, I'm guessing it will get GCed, and with it, all of its listeners.